File: back-end-component/REST/genetic-algorithm/src/main/java/com/mathapp/geneticalgorithm/model/GeneticAlgorithm.java
package com.mathapp.geneticalgorithm.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public class GeneticAlgorithm {
    // The size of the population
    private int populationSize;
    // The mutation rate
    private double mutationRate;
    // The crossover rate
    private double crossoverRate;
    // The tournament size for selection
    private int tournamentSize;
    // The target fitness score
    private double targetFitness;
    // The list of chromosomes in the population
    private List<Chromosome> population;

    /**
     * Constructor for creating a new GeneticAlgorithm object
     *
     * @param populationSize the size of the population
     * @param mutationRate   the mutation rate
     * @param crossoverRate  the crossover rate
     * @param tournamentSize the tournament size for selection
     * @param targetFitness  the target fitness score
     */
    public GeneticAlgorithm(int populationSize, double mutationRate, double crossoverRate, int tournamentSize, double targetFitness) {
        this.populationSize = populationSize;
        this.mutationRate = mutationRate;
        this.crossoverRate = crossoverRate;
        this.tournamentSize = tournamentSize;
        this.targetFitness = targetFitness;
        this.population = new ArrayList<>();
    }

    /**
     * Initializes the population with random chromosomes
     *
     * @param chromosomeLength the length of each chromosome
     */
    public void initializePopulation(int chromosomeLength) {
        for (int i = 0; i < populationSize; i++) {
            String genes = generateRandomGenes(chromosomeLength);
            double fitness = calculateFitness(genes);
            population.add(new Chromosome(genes, fitness));
        }
    }

    /**
     * Generates a random string of genes for a chromosome
     *
     * @param chromosomeLength the length of the chromosome
     * @return a string of random genes
     */
    private String generateRandomGenes(int chromosomeLength) {
        StringBuilder sb = new StringBuilder();
        Random rand = new Random();
        for (int i = 0; i < chromosomeLength; i++) {
            char gene = (char) (rand.nextInt(26) + 'a');
            sb.append(gene);
        }
        return sb.toString();
    }

    /**
     * Calculates the fitness score for a given set of genes
     *
     * @param genes the genes to evaluate
     * @return the fitness score
     */
    private double calculateFitness(String genes) {
        // Calculate the fitness score here based on your specific problem domain
        return 0.0;
    }

    /**
     * Selects a subset of the population for breeding using tournament selection
     *
     * @return a list of chromosomes selected for breeding
     */
    private List<Chromosome> tournamentSelection() {
        List<Chromosome> tournament = new ArrayList<>();
        Random rand = new Random();
        for (int i = 0; i < tournamentSize; i++) {
            int index = rand.nextInt(populationSize);
            tournament.add(population.get(index));
        }
        Collections.sort(tournament, Collections.reverseOrder());
        return tournament.subList(0, 2);
    }

    /**
     * Breeds two parent chromosomes to create two child chromosomes using single-point crossover
     *
     * @param parent1 the first parent chromosome
     * @param parent2 the second parent chromosome
     * @return a list of the resulting child chromosomes
     */
    private List<Chromosome> crossover(Chromosome parent1, Chromosome parent2) {
        List<Chromosome> children = new ArrayList<>();
        Random rand = new Random();
        if (rand.nextDouble() < crossoverRate) {
            int crossoverPoint = rand.nextInt(parent1.getGenes().length());
            String child1Genes = parent1.getGenes().substring(0, crossoverPoint) + parent2.getGenes().substring(crossoverPoint);
            String child2Genes = parent2.getGenes().substring(0, crossoverPoint) + parent1.getGenes().substring(crossoverPoint);
            double child1Fitness = calculateFitness(child1Genes);
            double child2Fitness = calculateFitness(child2Genes);
            children.add(new Chromosome(child1Genes, child1Fitness));
            children.add(new Chromosome(child2Genes, child2Fitness));
        } else {
            children.add(parent1);
            children.add(parent2);
        }
        return children;
    }

    /**
     * Mutates a given chromosome by randomly flipping some of its genes
     *
     * @param chromosome the chromosome to mutate
     * @return the mutated chromosome
     */
    private Chromosome mutate(Chromosome chromosome) {
        StringBuilder sb = new StringBuilder(chromosome.getGenes());
        Random rand = new Random();
        for (int i = 0; i < sb.length(); i++) {
            if (rand.nextDouble() < mutationRate) {
                char gene = (char) (rand.nextInt(26) + 'a');
                sb.setCharAt(i, gene);
            }
        }
        double fitness = calculateFitness(sb.toString());
        return new Chromosome(sb.toString(), fitness);
    }

    /**
     * Evolves the population by selecting parents, breeding them, and mutating the resulting children
     */
    public void evolve() {
        List<Chromosome> newPopulation = new ArrayList<>();
        while (newPopulation.size() < populationSize) {
            List<Chromosome> parents = tournamentSelection();
            List<Chromosome> children = crossover(parents.get(0), parents.get(1));
            Chromosome child1 = mutate(children.get(0));
            Chromosome child2 = mutate(children.get(1));
            newPopulation.add(child1);
            newPopulation.add(child2);
        }
        Collections.sort(newPopulation, Collections.reverseOrder());
        population = newPopulation;
    }

    /**
     * Runs the genetic algorithm until a solution with the target fitness score is found or the maximum number of generations is reached
     *
     * @param maxGenerations the maximum number of generations to run the algorithm for
     * @return the best chromosome found during the evolution process
     */
    public Chromosome run(int maxGenerations) {
        int generation = 0;
        Chromosome bestChromosome = population.get(0);
        while (bestChromosome.getFitness() < targetFitness && generation < maxGenerations) {
            evolve();
            bestChromosome = population.get(0);
            System.out.println("Generation " + generation + ": " + bestChromosome);
            generation++;
        }
        return bestChromosome;
    }

    // a bunch of generated setters for easy initialization in the controller


    public void setMutationRate(double mutationRate) {
        this.mutationRate = mutationRate;
    }

    public void setCrossoverRate(double crossoverRate) {
        this.crossoverRate = crossoverRate;
    }

    public void setTargetFitness(double targetFitness) {
        this.targetFitness = targetFitness;
    }
}

File: front-end-component/remake/algorithms/GeneticAlgorithm.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Globalization;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;

namespace MathApp.algorithms
{
    internal class GeneticAlgorithm
    {
        private Random _random;

        public GeneticAlgorithm()
        {
            _random = new Random();
            
        }

        public Complex SolveEquation(string equation)
        {
            int populationSize = 100;
            int chromosomeLength = 10;
            double minGeneValue = -10.0;
            double maxGeneValue = 10.0;
            int maxGenerations = 1000;
            double targetFitness = 0.99;

            // Generate an initial population
            List<Chromosome> population = GenerateInitialPopulation(populationSize, chromosomeLength, minGeneValue, maxGeneValue);

            for (int generation = 0; generation < maxGenerations; generation++)
            {
                // Evaluate the fitness of each chromosome in the population
                foreach (Chromosome chromosome in population)
                {
                    chromosome.Fitness = CalculateFitness(chromosome, equation);
                }

                // Check if the target fitness has been reached
                double bestFitness = population.Max(c => c.Fitness);
                if (bestFitness >= targetFitness)
                {
                    // Return the solution as a complex number
                    Chromosome bestChromosome = population.OrderByDescending(c => c.Fitness).First();
                    double[] solutionValues = bestChromosome.Genes;
                    Complex solution = new Complex(solutionValues[0], solutionValues[1]);
                    return solution;
                }

                // Evolve the population for the next generation
                population = EvolvePopulation(population);
            }

            // Return the best solution found as a complex number
            Chromosome bestChromosomeOverall = population.OrderByDescending(c => c.Fitness).First();
            double[] overallSolutionValues = bestChromosomeOverall.Genes;
            Complex overallSolution = new Complex(overallSolutionValues[0], overallSolutionValues[1]);
            return overallSolution;
        }


        private List<Chromosome> GenerateInitialPopulation(int populationSize, int chromosomeLength, double minGeneValue, double maxGeneValue)
        {
            List<Chromosome> population = new List<Chromosome>();

            for (int i = 0; i < populationSize; i++)
            {
                Chromosome chromosome = new Chromosome();

                for (int j = 0; j < chromosomeLength; j++)
                {
                    chromosome.Genes[j] = _random.NextDouble() * (maxGeneValue - minGeneValue) + minGeneValue;
                }

                population.Add(chromosome);
            }

            return population;
        }



        private List<Chromosome> EvolvePopulation(List<Chromosome> population)
        {
            List<Chromosome> offspringPopulation = new List<Chromosome>();

            // Perform tournament selection to select the parents for the crossover operation
            for (int i = 0; i < population.Count; i++)
            {
                Chromosome parent1 = TournamentSelection(population);
                Chromosome parent2 = TournamentSelection(population);

                // Perform crossover between the parents to create a new child chromosome
                Chromosome child = Crossover(parent1, parent2);

                // Perform mutation on the child chromosome
                Mutate(child);

                // Add the child chromosome to the offspring population
                offspringPopulation.Add(child);
            }

            // Replace the current population with the offspring population
            return offspringPopulation;
        }

        private Chromosome TournamentSelection(List<Chromosome> population)
        {
            int tournamentSize = 2;
            List<Chromosome> tournament = new List<Chromosome>();

            // Select a random subset of chromosomes from the population
            for (int i = 0; i < tournamentSize; i++)
            {
                int index = _random.Next(population.Count);
                tournament.Add(population[index]);
            }

            // Select the chromosome with the highest fitness from the tournament as the winner
            return tournament.OrderByDescending(c => c.Fitness).First();
        }


        private Chromosome Crossover(Chromosome parent1, Chromosome parent2)
        {
            // Perform single-point crossover between the parents to create a new child chromosome
            int crossoverPoint = _random.Next(parent1.Genes.Length);
            Chromosome child = new Chromosome();

            for (int i = 0; i < parent1.Genes.Length; i++)
            {
                if (i < crossoverPoint)
                {
                    child.Genes[i] = parent1.Genes[i];
                }
                else
                {
                    child.Genes[i] = parent2.Genes[i];
                }
            }

            return child;
        }

        private Chromosome Mutate(Chromosome chromosome)
        {
            // Perform uniform mutation on the chromosome
            double mutationProbability = 0.1;

            for (int i = 0; i < chromosome.Genes.Length; i++)
            {
                if (_random.NextDouble() < mutationProbability)
                {
                    chromosome.Genes[i] += _random.NextDouble() * 0.2 - 0.1;
                }
            }

            return chromosome;
        }

        private double CalculateFitness(Chromosome chromosome, string equation)
        {
            double totalFitness = 0.0;

            for (int i = 0; i < chromosome.Genes.Length; i++)
            {
                double geneValue = chromosome.Genes[i];
                string geneSymbol = "x" + (i + 1).ToString();

                // Replace the gene symbol with the gene value in the equation
                string evaluatedEquation = equation.Replace(geneSymbol, geneValue.ToString());

                try
                {
                    // Evaluate the equation and parse the result as a complex number
                    Complex solution = Complex.Parse(new DataTable().Compute(evaluatedEquation, null).ToString(), CultureInfo.InvariantCulture);

                    // Calculate the fitness based on the distance from the origin
                    double fitness = 1.0 / (1.0 + solution.Magnitude);
                    totalFitness += fitness;
                }
                catch (Exception ex)
                {
                    // Ignore exceptions caused by invalid equations
                    Console.WriteLine("Invalid equation: " + evaluatedEquation);
                }
            }

            // Calculate the average fitness for all the genes
            return totalFitness / chromosome.Genes.Length;
        }



        private class Chromosome
        {
            public double[] Genes { get; set; }
            public double Fitness { get; set; }

            public Chromosome()
            {
                Genes = new double[10];
                Fitness = 0.0;
            }
        }
    }

}
